crates/proto/src/xfer/mod.rs

```diff
@@ -207,47 +205,86 @@ impl OneshotDnsRequest {
     }
 }
 
-struct OneshotDnsResponse(oneshot::Sender<DnsResponseFuture>);
+struct OneshotDnsResponse(oneshot::Sender<DnsResponseStream>);
 
 impl OneshotDnsResponse {
-    fn send_response(self, serial_response: DnsResponseFuture) -> Result<(), DnsResponseFuture> {
+    fn send_response(self, serial_response: DnsResponseStream) -> Result<(), DnsResponseStream> {
         self.0.send(serial_response)
     }
 }
 
-/// A Future that wraps a oneshot::Receiver and resolves to the final value
-pub enum OneshotDnsResponseReceiver {
+/// A Stream that wraps a oneshot::Receiver<Stream> and resolves to items in the inner Stream
+pub enum DnsResponseReceiver {
     /// The receiver
-    Receiver(oneshot::Receiver<DnsResponseFuture>),
-    /// The future once received
-    Received(DnsResponseFuture),
+    Receiver(oneshot::Receiver<DnsResponseStream>),
+    /// The stream once received
+    Received(DnsResponseStream),
     /// Error during the send operation
     Err(Option<ProtoError>),
 }
 
-impl Future for OneshotDnsResponseReceiver {
-    type Output = Result<DnsResponse, ProtoError>;
+impl Stream for DnsResponseReceiver {
+    type Item = Result<DnsResponse, ProtoError>;
 
-    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         loop {
             *self = match *self.as_mut() {
-                OneshotDnsResponseReceiver::Receiver(ref mut receiver) => {
+                DnsResponseReceiver::Receiver(ref mut receiver) => {
                     let receiver = Pin::new(receiver);
                     let future = ready!(receiver
                         .poll(cx)
                         .map_err(|_| ProtoError::from("receiver was canceled")))?;
-                    OneshotDnsResponseReceiver::Received(future)
-                }
-                OneshotDnsResponseReceiver::Received(ref mut future) => {
-                    let future = Pin::new(future);
-                    return future.poll(cx);
+                    DnsResponseReceiver::Received(future)
                 }
-                OneshotDnsResponseReceiver::Err(ref mut err) => {
-                    return Poll::Ready(Err(err
-                        .take()
-                        .expect("futures should not be polled after complete")))
+                DnsResponseReceiver::Received(ref mut stream) => {
+                    return stream.poll_next_unpin(cx);
                 }
+                DnsResponseReceiver::Err(ref mut err) => return Poll::Ready(err.take().map(Err)),
             };
         }
     }
 }
+
+/// Helper trait to convert a Stream of dns response into a Future
+pub trait FirstAnswer<T, E: From<ProtoError>>: Stream<Item = Result<T, E>> + Unpin + Sized {
```

Comment:
This is a nice Future impl. I was thinking we'd need something like this. I like the impl.